Added FFT instance for Circat.Pair.

capn-freako opened the pull request:
In test/fft_test.hs, I've added an FFT instance for Circat.Pair, and am using it successfully, in the FFT instance for Circat.RTree. However, it's rather specific, which I don't like. (See comments in code.)

capn-freako committed:
Minor clean-up of fft_test.hs

Redundant bracket removal.

capn-freako committed:
Moved addPhase into FFT class definition.

- Tightened up the constraints on FFT class declaration.
- Pulled addPhase out of RTree instance and into class definition.
- Absorbed phasor into addPhase.

capn-freako committed:
Added FFT instance for Circat.Pair.

Am using it to implement FFT instance for Circat.RTree, but...
- My current instance for Pair assumes DIT. I don't like this.

conal commented on test/fft_test.hs:
```diff
@@ -39,31 +42,35 @@ import Circat.RTree (bottomSplit)
-class (LScan f) => FFT f where
-    fft  :: (RealFloat a) => f (Complex a) -> f (Complex a)
+--   The LScan constraint comes from the use of 'lproducts', in 'addPhase'.
+class (Applicative f, Foldable f, LScan f, Num (f a)) => FFT f a where
```
Why these parent classes?

capn-freako committed:
Moved many constraints from FFT class definition to Pair instance.

Thanks to Conal for pointing out that they needn't be in the FFT class
def.

capn-freako commented:
As per Conal, I've just moved several constraints out of the FFT class definition (where they weren't really needed) into the FFT instance for Circat.Pair (where they are, currently, needed).

capn-freako committed:
Copied lambda-ccc/test/fft_test.hs to fft-ccc.hs.

I decided to pull this file up to the top level of the TreeViz project,
since it really is intended just as a test vehicle for different
expressions of FFT, and not necessarily as compilable, via Conal's
machinery.

I've stated my intent, in the header of the lambda-ccc/test/fft_test.hs
file, asking that further general explorations in FFT expression be done
in fft-ccc.hs, instead.

capn-freako commented:
I've copied lambda-ccc/test/fft_test.hs to fft-ccc.hs (in my top-level TreeViz folder), and intend to do further general FFT exploration here. I'd like to use lambda-ccc/test/fft_test.hs strictly as a staging area for those FFT implementations, which we want to push through Conal's compiler. Thoughts?
